chirp-impl/src/main/java/sample/chirper/chirp/impl/JpaSessionImpl.java

```diff
@@ -65,31 +76,41 @@
-                    log.warn("Exception while initializing JPA EntityManagerFactory", e);
-                }
-                if (System.currentTimeMillis() - startTime <= INIT_TIMEOUT_MILLIS) {
-                    log.info("Will retry initializing JPA EntityManagerFactory in {} ms", INIT_RETRY_INTERVAL_MILLIS);
-                    try {
-                        Thread.sleep(INIT_RETRY_INTERVAL_MILLIS);
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                    }
-                }
-            }
-            log.error("Cancelled initializing JPA EntityManagerFactory after {} ms", INIT_TIMEOUT_MILLIS);
-            throw new CancellationException(
-                    "Cancelled initializing JPA EntityManagerFactory after " + INIT_TIMEOUT_MILLIS + " ms"
-            );
-        });
+    private CompletionStage<EntityManagerFactory> tryCreateEntityManagerFactory(FiniteDuration maxDuration,
+                                                                                FiniteDuration retryInterval) {
+        CompletionStage<EntityManagerFactory> result = executeInSlickContext(() ->
+                Persistence.createEntityManagerFactory(PERSISTENCE_UNIT_NAME)
+        );
+        if (maxDuration.toMillis() <= 0) {
+            // Duration is up, don't try and recover from failure
+            return result;
+        } else {
+            // This wrapping and then composing with identity logic is because
+            // CompletionStage has no equivalent to Scala's Future.recoverWith
+            return result.thenApply(CompletableFuture::completedFuture)
+                    .exceptionally(e -> retryCreateEntityManagerFactory(e, maxDuration, retryInterval))
+                    .thenCompose(Function.identity());
+        }
+    }
+
+    private CompletableFuture<EntityManagerFactory> retryCreateEntityManagerFactory(Throwable exception,
+                                                                                    FiniteDuration maxDuration,
+                                                                                    FiniteDuration retryInterval) {
+        log.warn("Exception while initializing JPA EntityManagerFactory", exception);
+        FiniteDuration delay = retryInterval.min(maxDuration);
+        log.info("Will retry initializing JPA EntityManagerFactory in {}", delay);
+        CompletableFuture<EntityManagerFactory> future = new CompletableFuture<>();
+        actorSystem.scheduler().scheduleOnce(
+                delay,
+                () -> tryCreateEntityManagerFactory(maxDuration.minus(delay), retryInterval)
+                        .thenAccept(future::complete)
+                        .exceptionally(ex -> {
+                            future.completeExceptionally(ex);
+                            return null;
+                        }),
+                actorSystem.dispatcher()
+        );
+        return future;
```

Comment:
Looks fine to me. One minor thing (and I know you based this off my initial logic), but I think the max duration waited ends up being the duration of waiting in the scheduler, and not the duration including the time it takes to create the entity manager factory. I think we probably need to create an end time at the start rather than work purely with durations.